R/expect_s3_class_linter.R

```diff
@@ -0,0 +1,108 @@
+#' Require usage of expect_s3_class()
+#'
+#' [testthat::expect_s3_class()] exists specifically for testing the class
+#'   of S3 objects. [testthat::expect_equal()], [testthat::expect_identical()],
+#'   and [testthat::expect_true()] can also be used for such tests,
+#'   but it is better to use the tailored function instead.
+#'
+#' @evalRd rd_tags("expect_s3_class_linter")
+#' @seealso [linters] for a complete list of linters available in lintr.
+#' @export
+expect_s3_class_linter <- function() {
+  Linter(function(source_file) {
+    if (length(source_file$parsed_content) == 0L) {
+      return(list())
+    }
+
+    xml <- source_file$xml_parsed_content
+
+    # (1) expect_{equal,identical}(class(x), C)
+    # (2) expect_true(is.<class>(x)) and expect_true(inherits(x, C))
+    is_class_call <- xp_text_in_table(c(is_s3_class_calls, "inherits"))
+    xpath <- glue::glue("//expr[
+      (
+        SYMBOL_FUNCTION_CALL[text() = 'expect_equal' or text() = 'expect_identical']
+        and following-sibling::expr[
+          expr[SYMBOL_FUNCTION_CALL[text() = 'class']]
+          and (position() = 1 or preceding-sibling::expr[STR_CONST])
```

Comment:
no... is it worth doing so? we're in the territory of extended logic for "non-canonical" usages already. Would mean growing the logic from

```
expr[STR_CONST]
```

to (roughly)

```
expr[STR_CONST or expr[SYMBOL_FUNCTION_CALL[text = 'c'] and not(expr[position() > 1 and not(STR_CONST)])]]
```

substantially more complicated because we need to do double-negation in the xpath.

Reply:
No, doesn't really seem worth it.

Reply:
definitely keep these cases in mind for the PR of yoda_test_linter, which looks specifically for the reversed usage that would be caught by the extended logic.

ideally users will have yoda_test_linter activated alongside whichever other expect_*_linter, which can first guide them to switch the arguments, and then to change the expectation.